R/dist_spec.R

```diff
@@ -549,75 +565,105 @@ print.dist_spec <- function(x, ...) {
-#' @param ... Additional arguments to pass to `{ggplot}`.
+#' @param samples Integer; Number of samples to generate for distributions
+#' with uncertain parameters (default: 50).
+#' @param res Numeric; Resolution of the PMF and CDF (default: 1, i.e. integer
+#'   discretisation).
+#' @param cumulative Logical; whether to plot the cumulative distribution in
+#'   addition to the probability mass function
+#' @param ... ignored
 #' @importFrom ggplot2 aes geom_col geom_step facet_wrap vars theme_bw
+#' @importFrom data.table data.table rbindlist
 #' @export
 #' @examples
-#' #' # A fixed lognormal distribution with mean 5 and sd 1.
+#' # A fixed lognormal distribution with mean 5 and sd 1.
 #' dist1 <- LogNormal(mean = 1.6, sd = 0.5, max = 20)
+#' # Plot discretised distribution with 1 day discretisation window
 #' plot(dist1)
+#' # Plot discretised distribution with 0.01 day discretisation window
+#' plot(dist1, res = 0.01, cumulative = FALSE)
 #'
 #' # An uncertain gamma distribution with mean 3 and sd 2
 #' dist2 <- Gamma(
 #'   mean = Normal(3, 0.5), sd = Normal(2, 0.5), max = 20
 #' )
 #' plot(dist2)
 #'
-#' # Multiple distributions
-#' plot(dist1 + dist2 + dist1)
-#'
-#' # A combination of the two fixed distributions
-#' plot(dist1 + dist1)
-plot.dist_spec <- function(x, ...) {
-  distribution <- cdf <- NULL
+#' # Multiple distributions with 0.1 discretisation window and do not plot the
+#' # cumulative distribution
+#' plot(dist1 + dist2, res = 0.1, cumulative = FALSE)
+plot.dist_spec <- function(x, samples = 50L, res = 1, cumulative = TRUE, ...) {
   # Get the PMF and CDF data
-  pmf_data <- data.frame(
-    value = numeric(), pmf = numeric(),
-    distribution = factor()
-  )
-  cdf_data <- data.frame(
-    value = numeric(), cdf = numeric(),
-    distribution = factor()
-  )
-  dist_sd <- sd_dist(x)
-  for (i in seq_along(x)) {
-    if (x[[i]]$distribution == "nonparametric") {
-      # Fixed distribution
+  pmf_data <- lapply(seq_len(ndist(x)), function(i) {
+    if (get_distribution(x, i) == "nonparametric") {
+      # nonparametric
       pmf <- get_pmf(x, i)
+      values <- seq_along(pmf) - 1
       dist_name <- paste0("Nonparametric", " (ID: ", i, ")")
+      pmf_dt <- data.table(
+        sample = 1, x = values, p = pmf, distribution = dist_name
+      )
     } else {
-      # Uncertain distribution
-      c_dist <- discretise(fix_dist(extract_single_dist(x, i)))
-      pmf <- get_pmf(c_dist)
+      # parametric
+      uncertain <- vapply(get_parameters(x, i), function(y) {
+        if (is.numeric(y)) return(FALSE)
+        sd_dist <- sd(y)
+        return(is.na(sd_dist) || sd_dist > 0)
+      }, logical(1))
+      if (!any(uncertain)) {
+        samples <- 1 ## only need 1 sample if fixed
+      }
+      dists <- lapply(seq_len(samples), function(y) {
+        fix_dist(extract_single_dist(x, i), strategy = "sample")
+      })
+      tolerance <- attr(x, "tolerance")
+      if (is.null(tolerance)) {
```

Comment:
See my comment about always setting `tolerance` in https://github.com/epiforecasts/EpiNow2/pull/724#issuecomment-2252676751. It'll cut out these logics.

Reply:
see main discussion - keeping this as is for now but to be reviewed if/when we move `<dist_spec>` into a separate package.